MOBT-2-Swift-Foundation.playground/Contents.swift

```diff
@@ -0,0 +1,332 @@
+}
+
+// MARK: - Protocols
+protocol CurrencyExchangeMethods {
+    func fetchJSON(retreiveDataKeys: @escaping DataKeys, retrieveDataValues: @escaping DataValues)
+}
+
+// MARK: - Classes
+final class CurrencyExchangeConverterService: CurrencyExchangeMethods {
+    // MARK: - Enums
+    private enum Const {
+        static let urlAPI: String = "https://open.er-api.com/v6/latest/USD"
+    }
+    
+    // MARK: - Properties
+    var currencyCode: [String] = []
+    var values: [Double] = []
+    
+    // MARK: - Methods
+    func fetchJSON(retreiveDataKeys: @escaping DataKeys, retrieveDataValues: @escaping DataValues) {
+        guard let url = URL(string: Const.urlAPI) else { return }
+        URLSession.shared.dataTask(with: url) { (data, response, error) in
+            if error != nil { return }
+            guard let data = data else { return }
+            do {
+                let results = try JSONDecoder().decode(Currency.self, from: data)
+                self.currencyCode.append(contentsOf: results.rates.keys)
+                self.values.append(contentsOf: results.rates.values)
+                retreiveDataKeys(self.currencyCode)
+                retrieveDataValues(self.values)
+            } catch {
+                print(error)
+            }
+        }.resume()
+    }
+}
+
+final class CurrencyExchangeLocalData {
+    var currencyExchangeService: CurrencyExchangeConverterService?
+    
+    init(currencyExchangeService: CurrencyExchangeConverterService?) {
+        self.currencyExchangeService = currencyExchangeService
+    }
+    
+    func setDataToUserDefaults() {
+        currencyExchangeService?.fetchJSON(retreiveDataKeys: { keys in
+            UserDefaults.standard.set(keys, forKey: Keys.mostRecentExchangeRatesKeys)
+        }, retrieveDataValues: { values in
+            UserDefaults.standard.set(values, forKey: Keys.mostRecentExchangeRatesValues)
+        })
+    }
+    
+    func retrieveDataFromUserDefaults() {
+        let currencyKeys = UserDefaults.standard.stringArray(forKey: Keys.mostRecentExchangeRatesKeys)
+        let currencyValues = UserDefaults.standard.object(forKey: Keys.mostRecentExchangeRatesValues)
+        print(currencyKeys ?? "No such keys in User Defaults")
+        print(currencyValues ?? "No such values in User Defaults")
+    }
+}
+
+let currencyExchangeService = CurrencyExchangeConverterService()
+let currencyExchangeLocalData = CurrencyExchangeLocalData(currencyExchangeService: currencyExchangeService)
+currencyExchangeLocalData.setDataToUserDefaults()
+currencyExchangeLocalData.retrieveDataFromUserDefaults()
+
+///When we have an object from the backend, we have a possibility to create MVVM-C architecture and use data to display on UI part
+///This is example of MVVM-C structure of our project
+
+///Base coordinator conforms to different coordinators to use methods of our currency exchanges
+class BaseCoordinator {
+    private(set) weak var rootController: UINavigationController?
```

Comment:
Why this property is optional if the initializer argument isn't? Looks like a development mistake